idp/authmiddleware/authmiddleware.go

```
@@ -0,0 +1,202 @@
+// If the user is logged in successfully information about the user (principal) and the authSession can be
+// taken from the context.
+// The parameter allowExternalValidation determines if the handler accepts external users. External users
+// are those who have been successfully authenticated by an external identity provider such as Google, but
+// have NOT been added to the pool of known users of this particular d.velop cloud tenant so far.
+// USE THIS FEATURE WITH CAUTION. You don't know much about external users and should restrict the rights
+// of external users to a minimum if you must allow access for external users at all. If external users are
+// enabled (allowExternalValidation is true) the principal struct representing an external user doesn't
+// provide any information apart from the e-mail address and a the reserved group ID '3E093BE5-CCCE-435D-99F8-544656B98681'
+// which marks the user as an external user which is unknown to the system. This group ID can be used to
+// distinguish external from internal users.
+// If you are unsure, you should set allowExternalValidation to false, as you usually don't want external users to access your app.
+//
+// Example:
+//	func main() {
+//		idpClient := &idp.Client{}
+//		authenticate := authmiddleware.Authenticate(idpClient, tenant.SystemBaseUriFromCtx, tenant.IdFromCtx, false, logerror, loginfo)
+//		authenticateExternal := authmiddleware.Authenticate(idpClient, tenant.SystemBaseUriFromCtx, tenant.IdFromCtx, true, logerror, loginfo)
+//		mux := http.NewServeMux()
+//		mux.Handle("/hello", authenticate(helloHandler()))
+//		mux.Handle("/resource", authenticate(resourceHandler()))
+//		mux.Handle("/resource4ExternalUsers", authenticateExternal(resource4ExternalUsersHandler()))
+//	}
+//
+//	func helloHandler() http.Handler {
+//		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+//			// get user from context
+//			principal,_ := authmiddleware.PrincipalFromCtx(r.Context())
```

Maybe it's a good idea to handle and not ignore the error in this example?

---

Done.